File: megumi/src/main/java/com/erisu/cloud/megumi/message/MessageUtil.java
package com.erisu.cloud.megumi.message;

import com.erisu.cloud.megumi.util.RedisUtil;
import net.mamoe.mirai.Bot;
import net.mamoe.mirai.contact.Group;
import net.mamoe.mirai.message.data.Message;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.scheduling.annotation.Async;
import org.springframework.stereotype.Component;

import javax.annotation.Resource;
import java.util.concurrent.TimeUnit;

/**
 * @Description send message
 * @Author alice
 * @Date 2021/1/13 16:54
 **/
@Component
public class MessageUtil {
    @Value("${qq.username}")
    private long username;
    @Resource
    private RedisUtil redisUtil;

    @Async
    public void sendAsyncMessage(Group group, Message message, Integer time) throws InterruptedException {
        TimeUnit.SECONDS.sleep(time);
        Bot.getInstance(username).getGroup(group.getId()).sendMessage(message);
    }

    /**
     * @param redisKey 基于redis决定任务是否发送（一般场景） 后续考虑如何更轻便 --todo 建议想好了重写
     * @param group
     * @param message
     * @param time     睡眠秒数
     * @throws InterruptedException
     */
    @Async
    public void sendAsyncMessageAwait(String redisKey, Group group, Message message, Integer time) throws InterruptedException {
        TimeUnit.SECONDS.sleep(time);
        if (!redisUtil.hasKey(redisKey)) {
            Bot.getInstance(username).getGroup(group.getId()).sendMessage(message);
        } else {
            redisUtil.delete(redisKey);
        }
    }
}

File: megumi/src/main/java/com/erisu/cloud/megumi/util/MessageUtil.kt
package com.erisu.cloud.megumi.util

import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.future.future
import net.mamoe.mirai.contact.Group
import net.mamoe.mirai.message.data.Image
import net.mamoe.mirai.message.data.Message
import net.mamoe.mirai.message.data.Voice
import net.mamoe.mirai.message.data.messageChainOf
import net.mamoe.mirai.utils.ExternalResource.Companion.toExternalResource
import java.io.File
import java.io.InputStream
import java.util.concurrent.CompletableFuture

/**
 *@Description 消息处理器
 *@Author alice
 *@Date 2021/6/22 14:37
 **/
object MessageUtil {
    fun message(vararg message: Message): Message {
        return messageChainOf(*message)
    }


    fun generateImageAsync(group: Group, file: File, delete: Boolean): CompletableFuture<Image> =
        GlobalScope.future { generateImage(group, file, delete) }

    fun generateImageAsync(group: Group, inputStream: InputStream): CompletableFuture<Image> =
        GlobalScope.future { generateImage(group, inputStream) }

    suspend fun generateImage(group: Group, file: File, delete: Boolean): Image {
        val externalResource = file.toExternalResource()
        val image: Image = group.uploadImage(externalResource)
        externalResource.close()
        if (delete) file.delete()
        return image
    }

    suspend fun generateAudio(group: Group, file: File, delete: Boolean): Voice {
        val externalResource = file.toExternalResource()
        val voice: Voice = group.uploadVoice(externalResource)
        externalResource.close()
        if (delete) file.delete()
        return voice
    }

    suspend fun generateAudio(group: Group, inputStream: InputStream): Voice {
        val externalResource = inputStream.toExternalResource()
        val voice: Voice = group.uploadVoice(externalResource)
        externalResource.close()
        return voice
    }

    suspend fun generateImage(group: Group, inputStream: InputStream): Image {
        val externalResource = inputStream.toExternalResource()
        val image: Image = group.uploadImage(externalResource)
        externalResource.close()
        inputStream.close()
        return image
    }


}